main.vala

```diff
@@ -639,291 +514,1289 @@ class Vls.Server {
+                }
             }
+        }
 
+        foreach (var entry in docs_not_published.entries) {
             try {
-                client.send_notification ("textDocument/publishDiagnostics", buildDict (
-                    uri: new Variant.string (uri),
-                    diagnostics: result
-                ));
+                client.send_notification (
+                    "textDocument/publishDiagnostics",
+                    buildDict (
+                        uri: new Variant.string (entry.key),
+                        diagnostics: new Variant.array (VariantType.VARIANT, new Variant[]{})
+                    ));
             } catch (Error e) {
-                debug (@"publishDiagnostics: failed to notify client: $(e.message)");
-                continue;
+                debug (@"[publishDiagnostics] failed to publish empty diags for $(entry.key): $(e.message)");
             }
+        }
+    }
+
+    Vala.CodeNode get_best (FindSymbol fs, TextDocument file) {
+        Vala.CodeNode? best = null;
 
-            debug (@"textDocument/publishDiagnostics: $uri");
+        foreach (var node in fs.result) {
+            if (best == null) {
+                best = node;
+            } else {
+                var best_begin = new Position.from_libvala (best.source_reference.begin);
+                var best_end = new Position.from_libvala (best.source_reference.end);
+                var node_begin = new Position.from_libvala (node.source_reference.begin);
+                var node_end = new Position.from_libvala (node.source_reference.end);
+
+                if (best_begin.compare (node_begin) <= 0 && node_end.compare (best_end) <= 0 &&
+                    // don't get implicit `this` accesses
+                    !(best.source_reference.begin.column == node.source_reference.begin.column &&
+                        node.source_reference.end.column == best.source_reference.end.column &&
+                        node is Vala.MemberAccess && 
+                        ((Vala.MemberAccess)node).member_name == "this" &&
+                        ((Vala.MemberAccess)node).inner == null))
+                    best = node;
+            }
         }
+
+        assert (best != null);
+        var sr = best.source_reference;
+        var from = (long)Server.get_string_pos (file.content, sr.begin.line-1, sr.begin.column-1);
+        var to = (long)Server.get_string_pos (file.content, sr.end.line-1, sr.end.column);
+        string contents = file.content [from:to];
+        debug ("Got best node: %s @ %s = %s", best.type_name, sr.to_string(), contents);
+
+        return (!) best;
+    }
+
+    Vala.Scope get_current_scope (Vala.CodeNode code_node) {
+        Vala.Scope? best = null;
+
+        for (Vala.CodeNode? node = code_node; node != null; node = node.parent_node) {
+            if (node is Vala.Symbol) {
+                var sym = (Vala.Symbol) node;
```

Comment:
do we care about non-symbol scopes? e.g. inside an `if()`?

Reply:
`Vala.IfStatement`s have `Block`s, which are symbols that contain the scopes.